### PR TITLE
Remove legacy providers interface

### PR DESCRIPTION
**Pull request by mtreinish**
### Summary

This commit removes the legacy providers interface from terra. This
includes the BaseBackend, BaseJob, BaseProvider abstract classes and
everything that inherits from them in the tree. These classes were
deprecated as part of qiskit-terra 0.18.0 and have been supersceded by
the versioned provider interface and all providers should have migrated
by now to these new supported interface. This also means we no longer
support the qobj execution path and QuantumCircuit and Schedule objects
are used directly everywhere for running. In the future we can look at
deprecating and removing the assemble() function and qobj class in the
future as nothing in tree is using them directly. But before we can
deprecate these the IBM provider and aer will need to have native
implementations.

### Details and comments

### CODE REVIEW

File: qiskit/utils/backend_utils.py [hunk 1]
```diff
@@ -37,7 +37,7 @@ def __init__(self) -> None:
 def _get_backend_interface_version(backend):
     """Get the backend version int."""
     backend_interface_version = getattr(backend, "version", None)
-    # Handle deprecated BaseBackend based backends which have a version()
+    # Handle deprecated Backend based backends which have a version()
     # method
     if not isinstance(backend_interface_version, int):
         backend_interface_version = 0
```
Comment (jakelishman) [inline qiskit/utils/backend_utils.py:43]: I think these lines are redundant now?

**Review by jakelishman — COMMENTED** [review]
The one minor comment aside, this all looks like a very straightforwards removal (hopefully!), and it'll certainly be good to see the back of all this legacy code.

File: qiskit/utils/backend_utils.py (hunk 1 above)
Comment (mtreinish) [reply qiskit/utils/backend_utils.py:43]: Yeah they are, it should never be triggered now, I'll drop them too

Comment (mtreinish) [reply qiskit/utils/backend_utils.py]: Done in: https://github.com/Qiskit/qiskit-terra/pull/7886/commits/090ae177611bcfbb1628113ad1cc0c657463977d

**Review by jakelishman — DISMISSED** [review]

**Review by jakelishman — APPROVED** [review]

**Comment by jakelishman** [comment]
@mtreinish: looks like this had an unfortunate CI timeout, and now there's a merge conflict.

**Review by jakelishman — APPROVED** [review]

**Pull request merged**